lib/python/picongpu/plugins/plot_mpl/energy_histogram_visualizer.py

```diff
@@ -0,0 +1,131 @@
+"""
+This file is part of the PIConGPU.
+
+Copyright 2017-2018 PIConGPU contributors
+Authors: Sebastian Starke
+License: GPLv3+
+"""
+
+from picongpu.plugins.energy_histogram import EnergyHistogram
+import matplotlib.pyplot as plt
+
+
+class Visualizer:
+    """
+    Class for creation of histogram plots on a logscaled y-axis.
+    """
+
+    def __init__(self, run_directory):
+        """
+        Parameters
+        ----------
+        run_directory : string
+            path to the run directory of PIConGPU
+            (the path before ``simOutput/``)
+        """
+        if run_directory is None:
+            raise ValueError('The run_directory parameter can not be None!')
+
+        self.energy_histogram = EnergyHistogram(run_directory)
+        self.plt_obj = None
+
+    def visualize(self, iteration, ax, species='e', species_filter="all",
```

Comment:
would be cool if the argument order would be as in the data api: species, species_filter, iteration, ax.

of course we can also invert additional options and iterations in both, but that should happen at once to be consistent

Reply:
Hm, this might be a problem for us, since as of right now we are using the visualizers in the following way (as a callback function of a jupyter-slider widget):

```{python}
iteration = change["new"]
idx = 0
for plugin in self.plugin_list:
    # each plugin gets an axes object for plotting
    # the sim data resides in a subdir of the experiment called \
    # 'run'
    plugin.visualize(iteration,
        self.ax_lst[idx],
        **self.layout_config)
        idx += 1
```
and the `layout_config` contains entries the user specifies (e.g. the species, filter and momentum type). Only this order allowed us to use this uniform way of calling the visualizers since they expect different optional parameters besides the iteration and the ax.
And the PNG Visualizer has (so far, at least) no options for filter or species (since I don't know whether you have PNG files for something other than electrons). 
All these points call for a common visualizer base class, what do you think? Can we create a common interface for all kinds of visualzers?

Reply:
Sounds reasonable, yes.

Reply:
By the way, the PNG visualizer does all its reading on its own and has no reader class like the others...maybe that should be split as well...

Reply:
Do you have png visualization also for other species than electrons? Does species_filter make sense for png visualization as well?
Only then we could use a base class for visualizers with the signature
```python
class BaseVisualizer
    def visualize(self, species, species_filter, iteration, **kwargs)
```

Reply:
yes, [the PNG plugin](https://picongpu.readthedocs.io/en/latest/usage/plugins/png.html#usage-plugins-png) can be used for any species and potentially in the future also `species_filters` (currently always `all`).

Reply:
Please be aware that not all visualizers are species visualizers.

Our sources can be species, [(electro-magnetic) fields](https://picongpu.readthedocs.io/en/latest/usage/plugins/sliceFieldPrinter.html#usage-plugins-slicefieldprinter) or something totally different, e.g. a [scalar of total energy](https://picongpu.readthedocs.io/en/latest/usage/plugins/energyFields.html#usage-plugins-energyfields).

Reply:
I see... so what would you think could be an interface that is 
a) general enough and 
b) also reflects the order of arguments of the data readers?

Reply:
probably the only common argument is `ax`. Although one could also query things globally for a simulation, in our case `iteration` could also be useful (maybe supporting ranges as well).

Reply:
When you have a png file, like `e_png_yz_0.5_006600.png`, what does the 0.5 stand for and what would be good parameter names for data readers for the coord label (yz) and that parameter behind the 0.5?
I'm asking because I'm trying to build a png reader as well

Reply:
`xy` is the plane that is cut out and `0.5` is the relative offset ("slicePoint") in the third, here `z` "axis" (in the box limit: [0.:1.]):
https://picongpu.readthedocs.io/en/latest/usage/plugins/png.html#usage-plugins-png